kotlin-jvm/src/main/kotlin/playground/NativeDI.kt

```diff
@@ -0,0 +1,225 @@
+/**
+ * Resolver have provider function for construction of dependencies
+ */
+object Resolver {
+
+    fun provideViewModel(sampleRepository: SampleRepository): ViewModel {
+        return ViewModel(sampleRepository)
+    }
+
+    fun provideSampleRepository(localDataSource: DataSourceLocal, remoteDataSource: DataSourceRemote): SampleRepository {
+        return SampleRepositoryImpl(localDataSource, remoteDataSource)
+    }
+
+    fun provideLocalDataStore(localDB: LocalDatabaseService): DataSourceLocal {
+        return DataSourceLocalImpl(localDB)
+    }
+
+    fun provideRemoteDataStore(apiService: ApiService): DataSourceRemote {
+        return DataSourceRemoteImpl(apiService)
+    }
+
+    fun provideLocalDatabase() = object : LocalDatabaseService() {}
+
+    fun provideApiService() = object : ApiService() {}
+}
+
+/**
+ * Injector creates and according to the needs cache or creates dependencies
+ * lazy can be used for caching and functions can be user to return new instances
+ */
+object Injector {
+    // construct and cache remote source
+    val apiService by lazy { Resolver.provideApiService() }
+    val remoteSource by lazy { Resolver.provideRemoteDataStore(apiService) }
+
+    // construct and cache local source
+    val localDB by lazy { Resolver.provideLocalDatabase() }
+    val localSource by lazy { Resolver.provideLocalDataStore(localDB) }
+
+    // construct and cache sample Repository
+    val sampleRepository by lazy { Resolver.provideSampleRepository(localSource, remoteSource) }
+
+    // construct and cache View or you can use function to get new viewmodel per call
+    val viewModel by lazy { Resolver.provideViewModel(sampleRepository) }
+}
+
+
+class ViewModel(private val sampleRepo: SampleRepository) {
+    var error: (Exception) -> Unit = {}
+    var userCallback: (SampleUser?) -> Unit = {}
+
+    fun getUser() = runBlocking {
+        //get cache value
+        println("reading from cache")
+        var user: SampleUser? = sampleRepo.getCacheUser()
+        // get remote value
+        if (user == null) {
+            println("reading remote cache")
+            val result = sampleRepo.getRemoteUser()
+            user = when (result) {
+                is Result.Success -> result.value
+                is Result.Error -> {
+                    error.invoke(result.error)
+                    null
+                }
+            }
+        }
+        // get last local value
+        if (user == null) {
+            println("reading local cache")
+            val result = sampleRepo.getLocalUser()
+            user = when (result) {
+                is Result.Success -> result.value
+                is Result.Error -> {
+                    error.invoke(result.error)
+                    null
+                }
+            }
+        }
+        userCallback.invoke(user)
+    }
+}
+
+class SampleRepositoryImpl(
+        private val dataSourceLocal: DataSourceLocal,
+        private val dataSourceRemote: DataSourceRemote
```

Comment:
Please, fix indent to match Kotlin official style guide.

Reply:
Hi @LouisCAD  I have set my studio with kotlin official style guide, it's showing the same result in auto-indentation and also checking on google show the same, can you tell me how I need to be formatted? something like
```
private val dataSourceLocal     : DataSourceLocal,
private val dataSourceRemote : DataSourceRemote
```

Reply:
@ch8n It should have only 4 spaces indent. Are you sure you saved the settings? Updating the code style and reformatting code with the IDE should do.